yogadl/_core.py

```diff
@@ -124,3 +124,12 @@ def cacheable(self, dataset_id: str, dataset_version: str) -> Callable:
         amongst instances of Storage in different processes.
         """
         pass
+
+    @abc.abstractmethod
+    def set_tensorflow_session(self, tensorflow_session: tf.compat.v1.Session) -> None:
+        """
+        Stores a tensorflow session to use to read a tf.data.Dataset submittable.
+
+        Only used for TF1.
+        """
+        pass
```

Comment:
This is not an OK extension to the core interface.  We are trying to write this core interface as something which is generally useful in SysML contexts, but this a very Determined-specific workaround for a specific corner case.

I don't have any better ideas yet though.  I'll go play with the use cases and get back to you.

Reply:
Can you be more precise in what you mean by:

> generally useful in SysML contexts

context of people who use machine learning frameworks?

Reply:
Ok, well I was still thinking we could just fork the process to do the submission, but it didn't take very much googling to find that [tf doesn't support that](https://github.com/tensorflow/tensorflow/issues/8220#issuecomment-302826884).  Not sure what I expected honestly.

Next idea: the problem we are actually trying to fix is not that we need the same session, but that we need the session to use the same GPUs as the session we later train with, right?

When I suggested just pass the session as a way to accomplish that, it seemed like a good idea, but I don't like how it is playing out here.  I guess I was thinking about passing it as an optional argument to `@cacheable`.  I think that's a lot better than having a separate method that every Storage implementation will have to support named `set_tensorflow_session()` which is only even useful in a basically-deprecated library.

But the more I think about it, I think we should leave this detail out of the core interfaces entirely, and just set the GPUs that the session is allowed to use in our `Storage` implementation.  It's not really useful for us to put the actual session there, since the session isn't likely to be created by the time we need to instantiate the `Storage` class in our system, but setting the GPUs to use is actually something that has a chance of being useful in a non-TF1 use case.

Also, I actually think this is more of what you were originally suggesting, right?

Reply:
> just set the GPUs that the session is allowed to use in our Storage implementation

The one worry I have about making the the GPU id configurable is that it's a bit misleading since we are not actually using a GPU to do the preprocessing. Also I am not sure how well it will work if someone passes a GPU id and then uses `tf.distributed_strategies`. Obviously in this later case I think not passing a GPU id would also work but seems a bit fragile. My original idea was to pass a `tf.ConfigProto`. 

>  It's not really useful for us to put the actual session there, since the session isn't likely to be created by the time we need to instantiate the Storage class in our system

The session can definitely be created by the time we instantiate the storage class. 

If you want to avoid extending our core interface we could just do:

```
class LFSStorage(yogadl.Storage):
    """
    Storage for local file system (not NFS).
    """

    def __init__(self, configurations: LFSConfigurations, tf_session: Optional[tf.Session]):
```

Also can do same thing with GPU ids instead.

Reply:
> Also can do same thing with GPU ids instead
Yes, this is exactly what I am describing.

> The session can definitely be created by the time we instantiate the storage class

Ok, I guess this is true.  I had to talk to Brian about the ordering of `disable_eager_execution` in TFKerasTrial.

> The one worry I have about making the the GPU id configurable is that it's a bit misleading since we are not actually using a GPU to do the preprocessing

Good point.  Ok, I'm fine with putting the session in the LFSStorage section.

Reply:
> Ok, I guess this is true. I had to talk to Brian about the ordering of disable_eager_execution in TFKerasTrial.

don't think this is an issue, because eager execution should be `enabled/disabled` at the beginning of user code, and we configure `storage` after user code is imported.

Reply:
Wait one more thing here:  If we pass a `tf.session` here instead of GPU IDs, then we will have to have a bunch of different codepaths for how to create the storage object, since the session is created in the framework logic-level.

We would only have to create the Storage in one place if we used GPU IDs.